client-dotnet/src/Controller/Controllers/EmController.cs

```diff
@@ -0,0 +1,280 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Client.Util;
+using Controller.Api;
+using Controller.Problem;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Controller.Controllers {
+
+  public class EmController : ControllerBase {
+
+    private readonly SutController _sutController;
+    private static string _baseUrlOfSut;
+
+    /*
+     Keep track of all host:port clients connect so far.
+     This is the mainly done for debugging, to check that we are using
+     a single TCP connection, instead of creating new ones at each request.
+     
+     However, we want to check it only during testing
+     */
+    //TODO: check
+    private static readonly ICollection<string> connectedClientsSoFar = new SynchronizedCollection<string> ();
+
+    private static readonly SemaphoreLocker _locker = new SemaphoreLocker ();
+
+    //ATTENTION: The relative path is from the SUT
+    private static readonly string htmlWarning = System.IO.File.ReadAllText ("../../../../client-dotnet/src/Controller/Resources/warning.html");
+
+    private readonly object syncLock = new object ();
+
+    public EmController (SutController sutController) {
+
+      if (!sutController.Equals (null))
+        _sutController = sutController;
+      else
+        throw new NullReferenceException ("SutController shouldn't be null");
+    }
+
+    private bool TrackRequestSource (ConnectionInfo connectionInfo) {
+
+      string source = $"{connectionInfo.RemoteIpAddress}:{connectionInfo.RemotePort}";
+
+      connectedClientsSoFar.Add (source);
+
+      return true;
+    }
+
+    private void AssertTrackRequestSource (ConnectionInfo connectionInfo) {
+
+      var res = TrackRequestSource (connectionInfo);
+
+      if (!res) throw new InvalidOperationException ();
+    }
+
+    //Only used for debugging/testing
+    ///<summary>Returns host:port of all clients connected so far</summary>
+    public static HashSet<string> GetConnectedClientsSoFar () => connectedClientsSoFar.ToHashSet ();
```

Comment:
instead of `HashSet<string>`, shouldn't it be a `ISet<string>`? ie, type declaration for containers should use the interface, and not the concrete classes

Reply:
Changed to ISet<string>